src/rdkafka_partition.c

```diff
@@ -4479,3 +4499,149 @@ const char *rd_kafka_fetch_pos2str(const rd_kafka_fetch_pos_t fetchpos) {
+}
+
+/**
+ * @brief Calculates \p a - \p b
+ *        Ordered following \p a order. Elements are copied from \p a.
+ */
+static rd_kafka_topic_partition_list_t *
+rd_kafka_topic_partition_list_difference0(rd_kafka_topic_partition_list_t *a,
+                                          rd_kafka_topic_partition_list_t *b,
+                                          int(cmp)(const void *_a,
+                                                   const void *_b),
+                                          unsigned int(hash)(const void *_a)) {
+        rd_kafka_topic_partition_t *rktpar;
+        rd_kafka_topic_partition_list_t *ret =
+            rd_kafka_topic_partition_list_new(a->cnt);
+        map_toppar_void_t b_map =
+            RD_MAP_INITIALIZER(b->cnt, cmp, hash, NULL, NULL);
+        RD_KAFKA_TPLIST_FOREACH(rktpar, b) {
+                RD_MAP_SET(&b_map, rktpar, rktpar);
+        }
+        RD_KAFKA_TPLIST_FOREACH(rktpar, a) {
+                if ((RD_MAP_GET(&b_map, rktpar) != NULL) == 0) {
+                        rd_kafka_topic_partition_list_add_copy(ret, rktpar);
+                }
+        }
+        RD_MAP_DESTROY(&b_map);
+        return ret;
+}
+
+/**
+ * @brief Calculates \p a ∩ \p b.
+ *        Ordered following \p a order. Elements are copied from \p a.
+ */
+rd_kafka_topic_partition_list_t *
+rd_kafka_topic_partition_list_intersection(rd_kafka_topic_partition_list_t *a,
+                                           rd_kafka_topic_partition_list_t *b) {
+        return rd_kafka_topic_partition_list_intersection0(
+            a, b, rd_kafka_topic_partition_cmp, rd_kafka_topic_partition_hash);
+}
+
+/**
+ * @brief Calculates \p a - \p b.
+ *        Ordered following \p a order. Elements are copied from \p a.
+ */
+rd_kafka_topic_partition_list_t *
+rd_kafka_topic_partition_list_difference(rd_kafka_topic_partition_list_t *a,
+                                         rd_kafka_topic_partition_list_t *b) {
+        return rd_kafka_topic_partition_list_difference0(
+            a, b, rd_kafka_topic_partition_cmp, rd_kafka_topic_partition_hash);
+}
+
+/**
+ * @brief Calculates \p a ∪ \p b.
+ *        Ordered following \p a order for elements in \p a
+ *        and \p b order for elements only in \p b.
+ *        Elements are copied the same way.
+ */
+rd_kafka_topic_partition_list_t *
+rd_kafka_topic_partition_list_union(rd_kafka_topic_partition_list_t *a,
+                                    rd_kafka_topic_partition_list_t *b) {
+        rd_kafka_topic_partition_list_t *ret =
+            rd_kafka_topic_partition_list_copy(a);
+        rd_kafka_topic_partition_list_grow(ret, b->cnt);
+        rd_kafka_topic_partition_list_t *b_minus_a =
+            rd_kafka_topic_partition_list_difference0(
+                b, a, rd_kafka_topic_partition_cmp,
+                rd_kafka_topic_partition_hash);
+        rd_kafka_topic_partition_list_add_list(ret, b_minus_a);
+        rd_kafka_topic_partition_list_destroy(b_minus_a);
+        return ret;
```

Comment:
```suggestion
        rd_kafka_topic_partition_list_t *b_minus_a =
            rd_kafka_topic_partition_list_difference0(
                b, a, rd_kafka_topic_partition_cmp,
                rd_kafka_topic_partition_hash);
        rd_kafka_topic_partition_list_t *ret =
            rd_kafka_topic_partition_list_new(a->cnt + b_minus_a->cnt);
        rd_kafka_topic_partition_list_add_copy(ret, a);
        rd_kafka_topic_partition_list_add_copy(ret, b_minus_a);
        rd_kafka_topic_partition_list_destroy(b_minus_a);
        return ret;
```

Reply:
Nice improvement, I've just kept `rd_kafka_topic_partition_list_add_list` because `rd_kafka_topic_partition_list_add_copy` adds a single partition not a whole list, but internally it uses `rd_kafka_topic_partition_list_add_copy`